### ISSUE TITLE
Add synthetic model test framework for rapid pipeline validation and testing of model's ability to learn

### ISSUE THREAD

**Issue by AoifeHughes** [alan-turing-institute/ice-station-zebra#367]
As discussed during co-working session on 9/7/2026. 

I think it would be useful to have some semi-automated testing and validation of model pipelines that use very simple synthetic data. My current thoughts are: 

1. Generate a binary image with a circle drawn in 1's
2. Perform some basic matrix shifting to simulate 'moving' of ice and create a set of test/train/val images
3. Use existing configs to pass this simulated data 
4. Verify outputs are as expected.

If a model couldn't learn a simple linear movement then it would flag and save running larger costly training. Having this simple and efficient enough would be good for CI reasons too to ensure we don't make breaking changes to model logic.

**Comment by AoifeHughes** [comment]
Depending on how this initial stage works out, I'd be interested in exploring:
1. More complex transformations, e.g., using opening/closing morphological operations, multiple patches.
2. Consider if it's viable to emulate other variables based on shape, research assumptions on effects e.g., 'surface temp does X to ice', use these assumptions to prototype additional complexity 
3. Building upon this might help with #269 and guide any assumptions

These last 2 points may just be me trying to [turn everything into a sphere](https://en.wikipedia.org/wiki/Spherical_cow) for easier modelling and not actually useful. But my instincts are that the data can be very noisy and 27-ish channels of noise is creating a very complex problem on top of developing a modelling framework, and anything to try and reduce complexity and test hypotheses against would be useful. 

On the hypothesis testing side I think this relates to point 2. In that if we say that we hypothesize that X variable is useful in predicting Y, yet a dummy model with simple data can't show that, then it might give us cause+ a playground in which to explore and actually come up with more novel hypotheses e.g., 'X variable is useful in predicting Y, but only when Z variable is also included in the model'. 

Again, to stress, I have an inkling that the basic version of this could be prototyped and ready in a few days, the more ambitious stuff may actually be more complexity and be more of a timesink than just running small models with real data. 

Happy for others' thoughts, also happy to be told I'm going off on a tangent and to reel it in.

**Comment by IFenton** [comment]
I definitely like the your simple idea. Sounds very useful, and I've been thinking for a while we should have something in the CI that can test the full pipeline logic. 

I'd have to think a bit more about the extensions. At the moment I'm not sure whether we can expect the relationships between say temperature and sea ice to be simple enough to model like this, but it might be worth exploring.

**Comment by AoifeHughes** [comment]
> I'd have to think a bit more about the extensions. At the moment I'm not sure whether we can expect the relationships between say temperature and sea ice to be simple enough to model like this, but it might be worth exploring.

I agree, it's not super straight-forward or simple. I'd be afraid that by using all the parameters that we have we're by proxy saying that we hypothesize a relationship between all of these things and predicting change in ice concentration. So I'd be keen to try and find a way to address that. Even if it means doing it via real data ultimately.

**Comment by AoifeHughes** [comment]
Started testing / prototyping in: https://github.com/alan-turing-institute/icenet-mp/tree/ah/model-testing. 

And I'm still working out if Im posting to wandb correctly: https://wandb.ai/turing-seaice/train/runs/2ybojfjg?nw=nwuserahug

**Comment by AoifeHughes** [comment]
Just copying over some images that I'd also shared on slack on some of the testing I've done with this

Using a UNet base model flagged some issues, tried various things such as lifting EMA and it just led to instability.

<img width="3664" height="1770" alt="Image" src="https://github.com/user-attachments/assets/40c9d464-ca66-484b-8643-27f4f73aac1d" />

I used the vit cnn model through this and it fit better,

<img width="3664" height="1770" alt="Image" src="https://github.com/user-attachments/assets/cc05103b-217d-4448-991a-0971bc1d1423" />

reduced the patch size and it improved too

<img width="3664" height="1770" alt="Image" src="https://github.com/user-attachments/assets/684bca2b-75df-498d-bb21-b0f5dc5ff9fc" />

and also with a down scaled 32x32 model for testing, it trained in a fraction of the time and shows the model setup works

<img width="3664" height="1770" alt="Image" src="https://github.com/user-attachments/assets/f7292448-d1be-491b-9dd2-c489e896908e" />

I think these examples somewhat illustrate why I think something like this could be useful for testing a model's strengths/weaknesses in a controlled way, and can more easily shift through training data e.g., looking at the unet outputs indicates there may be a BatchNorm train/eval mismatch, and with the cnn vit it allowed for quick testing of patch sizes for accuracy increases.

**Comment by AoifeHughes** [comment]
@jemrobinson made a great point we should use something along the lines of [opening/closing operations](https://en.wikipedia.org/wiki/Opening_(morphology)). 

TODO: 
- [x] Add this
- [ ] Test alone
- [ ] Test combining this with the moving model (maybe slow down the moving)

**Comment by AoifeHughes** [comment]
@Shaerdan if you're interested I just pushed https://github.com/alan-turing-institute/icenet-mp/blob/ah/model-testing/docs/src/how-to/synthetic-check.md - if you've any issues let me know. 

I need to wrestle with some formatting / lint things as well as updating the branch to be based off main's recent changes, but I think it should work in that state. I'll be pushing a few updates through the day though. Shout if anything throws issues if you do get around to trying it out

**Comment by AoifeHughes** [comment]
CI for this seems to be working:

- [Run here](https://github.com/alan-turing-institute/icenet-mp/actions/runs/29242108718/job/86790325309?pr=371)
- [Output here](https://github.com/alan-turing-institute/icenet-mp/actions/runs/29242108718/artifacts/8275888516)

Took ~2 mins to run on GH. Could probably go smaller and push it a little more and/or reduce epochs and really slim it down to < 1 minute per run. 

Metrics probably need to be more sensible, but I think it's fine for now. 

<img width="600" height="400" alt="Image" src="https://github.com/user-attachments/assets/68b7d484-e3aa-4c97-9eb8-708675aba025" />

<img width="3665" height="1770" alt="Image" src="https://github.com/user-attachments/assets/7b01782c-157a-4510-81a9-b33bd0ff3276" />

Like I'd mentioned earlier, there's probably a nice way to detect 'expected' vs. 'broken' states. Given this one it's clear that the prediction is in the right area etc. it's fair to assume the model isn't fundamentally broken and that the pipeline isn't failing

**Comment by AoifeHughes** [comment]
Thinking on it more, the opening/closing method could be interesting to apply to a 2 channel input model for testing, have the size of it be proportional to a second channel (doesn't matter what it's called, but for example could use degrees celsius relative values), then bundle the values together.
